Adds an f-plane implementation with arbitrary direction of rotation

tomchor opened the pull request:
Closes https://github.com/CliMA/Oceananigans.jl/issues/1372

~For now I just wanted to save up the calculations. Once we decide how to implement the feature I'll finish the PR.~

tomchor commented:
From #1372:

> On the other hand, it's a bit confusing that all Coriolis forces ultimately represent a "background rotation rate"...

Yeah I agree this might make it a bit confusing. For now I didn't delete anything and created `GeneralFPlane` as a placeholder. Would it be too bad to just have `FPlane` and let it default to `rotation_axis=ZDIrection()`? I think this is simple and would do the job well enough.

tomchor commented:
On a refactoring note, currently `ZDirection` is defined here https://github.com/CliMA/Oceananigans.jl/blob/fafdf3c7caa6aa1431b9d6437036794ec5dcb8fc/src/BuoyancyModels/buoyancy.jl#L6

Which is after `Coriolis/coriolis.jl` is compiled. I think it makes more sense to move this definition to `Grids`. Everyone okay with that? That way any other module can use it.

I also think would be useful to use something like `validate_vertical_unit_vector()` to validate the rotation axis: https://github.com/CliMA/Oceananigans.jl/blob/fafdf3c7caa6aa1431b9d6437036794ec5dcb8fc/src/BuoyancyModels/buoyancy.jl#L37

So I was thinking of moving this to `Utils` (or maybe even `Grids`?) and renaming it to `validate_unit_vector()`.

Since this would be refactoring code, I'll wait for some feedback before doing these modifications.

francispoulin commented:
Thanks for creating this @tomchor , I think this is a neat idea.

To help me think about how this should look, could you help me find an example you want want to do this?  

I can imagine that maybe you would want a general fplane and general nontraditional fplane as well.  I guess it depends on the physical set up.

tomchor commented:
> To help me think about how this should look, could you help me find an example you want want to do this?

Sure thing! The example I'm developing this for specifically is a tilted bottom boundary layer. (https://github.com/CliMA/Oceananigans.jl/pull/1498)

Over there I'm tilting the domain by tilting the gravity vector, really. However, to "properly" tilt the domain I'd need to tilt Coriolis too. In this case I'd need a x-component to `f`, which isn't available through `FPlane` or `NonTraditionalFPlane`. Thus, with this addition I'd be able to write something like

```julia
const θ_rad = 0.2 # radians
const g̃ = (sin(θ_rad), 0, cos(θ_rad)) # domain tilt

buoyancy = Buoyancy(model=BuoyancyTracer(), vertical_unit_vector=g̃) # Tilt gravity vector
coriolis = GeneralFPlane(coriolis_frequency=1e-4, rotation_axes=g̃) # Tilt rotation vector
```

tomchor commented:
> I can imagine that maybe you would want a general fplane and general nontraditional fplane as well. I guess it depends on the physical set up.

I'm actually not sure. All that `NonTraditionalFPlane` does is to tilt the rotation axis of an f-plane based on latitude. So it just projects some component of `f` onto the `y` direction. `GeneralFPlane` (or whatever we end up calling it) can tilt the rotation axis in any arbitrary direction, so it can already do what `NonTraditionalFPlane` does (and more). So really (imho) I don't see much use for `NonTraditionalFPlane` after this gets implemented, except maybe as a convenience function.

glwagner commented:
> > I can imagine that maybe you would want a general fplane and general nontraditional fplane as well. I guess it depends on the physical set up.
> 
> I'm actually not sure. All that `NonTraditionalFPlane` does is to tilt the rotation axis of an f-plane based on latitude. So it just projects some component of `f` onto the `y` direction. `GeneralFPlane` (or whatever we end up calling it) can tilt the rotation axis in any arbitrary direction, so it can already do what `NonTraditionalFPlane` does (and more). So really (imho) I don't see much use for `NonTraditionalFPlane` after this gets implemented, except maybe as a convenience function.

This should definitely replace `NonTraditionalFPlane`.

tomchor commented:
As of right now, as far as I can tell, `GeneralFPlane` is working and `NonTraditionalFPlane` got dumped. I'd like to get some feedback before I start changing the docs if that's okay.

The interface I implemented is a bit simpler than `NonTraditionalFPlane` but I think that's okay. I'd like some feedback there if possible. Arguments now are:

- `coriolis_frequency` (instead of simply `f`). Defaults to `2Ω`
- `rotation_axis` (defaults to `ZDirection`)
- `latitude` for convenience (which overwrites `rotation_axis`). I'm a bit unsure if I should keep this one. Maybe we should keep it simple since anyone can figure out `f` based on latitude easily.

Questions:

- Should we keep the original `FPlane`? I think we could scrap it and only keep this one (which would then be renamed `FPlane`)
- I didn't see any tests that `FPlane` or any other Coriolis implementation is dynamically correct. Should we implement a test to see if our rotation implementations are actually doing what they're supposed to?

francispoulin commented on src/BuoyancyModels/buoyancy.jl:
```diff
@@ -28,22 +29,10 @@ model = NonhydrostaticModel(grid=grid, buoyancy=buoyancy, tracers=:b)
 """
 function Buoyancy(; model, vertical_unit_vector=ZDirection())
-    vertical_unit_vector = validate_vertical_unit_vector(vertical_unit_vector)
+    vertical_unit_vector = validate_unit_vector(vertical_unit_vector)
```
makes sense to me to move `validate_unit_vector` elsewhere. Good idea.

tomchor replied:
Thanks :)

francispoulin commented on src/Coriolis/general_f_plane.jl:
```diff
@@ -0,0 +1,63 @@
+    GeneralFPlane{FT} <: AbstractRotation
+
+A Coriolis implementation that accounts for the locally vertical and
+both local horizontal components of the rotation vector. Traditionally
```
maybe add 'possibly` before both, as we might have 2, 1 or 0?

tomchor replied:
Done!

francispoulin commented on src/Coriolis/general_f_plane.jl:
```diff
@@ -0,0 +1,63 @@
+@inline fʸw_minus_fᶻv(i, j, k, grid, coriolis, U) =
+    coriolis.fy * ℑzᵃᵃᶜ(i, j, k, grid, U.w) - coriolis.fz * ℑyᵃᶜᵃ(i, j, k, grid, U.v)
+
+@inline fᶻu_minus_fˣw(i, j, k, grid, coriolis, U) =
```
I have never seen the `fx` compoments before and nice to see them for the first time!

tomchor replied:
Hopefully I did the math right :joy: 

But that's also part of the reason why I think we could test that Coriolis is doing what it's supposed to be doing

francispoulin replied:
I double checked and the equations look good to me.

francispoulin commented on src/Coriolis/general_f_plane.jl:
```diff
@@ -0,0 +1,63 @@
+@inline fˣv_minus_fʸu(i, j, k, grid, coriolis, U) =
+    coriolis.fx * ℑyᵃᶜᵃ(i, j, k, grid, U.v) - coriolis.fy * ℑxᶜᵃᵃ(i, j, k, grid, U.u)
+
+@inline x_f_cross_U(i, j, k, grid, coriolis::GeneralFPlane, U) = ℑxᶠᵃᵃ(i, j, k, grid, fʸw_minus_fᶻv, coriolis, U)
```
It seems there is some averaging going on when you compute, say `coriolis.fy`, but then you average again to compute `y_f_crosss_U`.  Is the averaging different?  This only happenes once so I don't think it's going to have any slow down at all but I'm just curious.

tomchor replied:
I'm not sure what averaging you're talking about. Are you talking about the use of `ℑxᶜᵃᵃ` operators?

francispoulin replied:
Yes, the appear both on the calculation of the first terms and the second terms as well.

tomchor replied:
I copied these from the previous implementation of `NonTraditionalFPlane`, but from what I understand `fˣv_minus_fʸu` interpolates `u` and `v` to `(Center, Center, Center)`, so that they can be later interpolated to `(Center, Center, Face)`. The reason is that this is going to be using in the prognostic equation for `w`, which exists in `(Center, Center, Face)` nodes. (And so on for the other functions.)

francispoulin commented on test/test_coriolis.jl:
```diff
@@ -12,15 +12,16 @@ function instantiate_fplane_2(FT)
-    coriolis = NonTraditionalFPlane(FT, fz=π, fy=π)
-    @test coriolis.fz == FT(π)
-    @test coriolis.fy == FT(π)
+    coriolis = GeneralFPlane(FT, coriolis_frequency=1, rotation_axis=[0, cosd(45), sind(45)])
```
After you do this for the `FPlane` are you thinking of doing the `BetaPlane` too?  I'm guessing it would be easier than this and might be nice to have.

tomchor replied:
I agree it would be easier to have, but I don't intend on doing this in this PR. I've also never run a simulation with the Beta plane and only study small scale turbulence (where the beta effect doesn't matter) so it might be best to leave that endeavor to someone who has more familiarity with it...

francispoulin replied:
I did not mean to suggest you do it here.  That would be better saved for a later PR, and should be revistied afterwards.  But it would make sense to be consistent across all of our planes, long term.

francispoulin commented on test/test_coriolis.jl:
```diff
@@ -12,15 +12,16 @@ function instantiate_fplane_2(FT)
+    @test coriolis.fz == FT(sind(45))
 end
 
 function instantiate_ntfplane_2(FT)
```
since `NontraditionalFPlane` is going to disappear, would we want to change the name of this function appropriately?  Say 'instantiate_general_fplane_2` or something?

glwagner replied:
Also lets definitely not break the convention of using verbose names!

tomchor replied:
Ah, good point. I missed that :)

francispoulin reviewed (approved):
Lots of nice work here @tomchor, thanks for putting this together!

francispoulin commented:
Thoughts on your questions:

- I can imagine some people would prefer to type `FPlane` instead fo `GeneralFPlane`.  Maybe we could keep the name of the former but get it to call the latter?  Bascially keep the name but none of the code, as having fewer functions seems like a good idea.
- We could look at a propagating internal wave and measure it's phase speed.  That would be one way to validate the Corioilis parameter, but maybe more complicated than other examples.

glwagner commented:
> Closes #1372
> 
> For now I just wanted to save up the calculations. Once we decide how to implement the feature I'll finish the PR.

I've been asked to review --- is this still the state of this PR?

tomchor commented:
> > Closes #1372
> > For now I just wanted to save up the calculations. Once we decide how to implement the feature I'll finish the PR.
> 
> I've been asked to review --- is this still the state of this PR?

No it's not. Thanks, I forgot to change that. 

It's now ready for review with the caveat that I'd like to wait until I get feedback on how it's working to change the docs if that's okay, as per [this comment](https://github.com/CliMA/Oceananigans.jl/pull/1892#issuecomment-887056354). I just wanna make sure you're all okay with the changes in functionality. But lmk if that's not okay and I'll change the docs too.

glwagner commented:
> As of right now, as far as I can tell, `GeneralFPlane` is working and `NonTraditionalFPlane` got dumped. I'd like to get some feedback before I start changing the docs if that's okay.
> 
> The interface I implemented is a bit simpler than `NonTraditionalFPlane` but I think that's okay. I'd like some feedback there if possible. Arguments now are:
> 
> * `coriolis_frequency` (instead of simply `f`). Defaults to `2Ω`
> * `rotation_axis` (defaults to `ZDirection`)
> * `latitude` for convenience (which overwrites `rotation_axis`). I'm a bit unsure if I should keep this one. Maybe we should keep it simple since anyone can figure out `f` based on latitude easily.
> 
> Questions:
> 
> * Should we keep the original `FPlane`? I think we could scrap it and only keep this one (which would then be renamed `FPlane`)
> * I didn't see any tests that `FPlane` or any other Coriolis implementation is dynamically correct. Should we implement a test to see if our rotation implementations are actually doing what they're supposed to?

For API this is what I suggest:

Three "modes":

1. A "primitive" mode whereby fx, fy, fz are all explicitly provided.
2. A "general rotation" mode whereby `rotation_axis` and `rotation_rate` are provided. `fx`, `fy`, and `fz` are then calculated as a convenience.
3. A "tangent plane" mode whereby `latitude` and `rotation_rate` are provided. This mode calculates `rotation_axis` based on `latitude` (and assuming that `y` is north-south).

The code might look something like

```julia
zero_if_nothing(f) = f
zero_if_nothing(::Nothing) = 0

function ConstantBackgroundRotation(FT=Float64; fx=nothing, fy=nothing, fz=nothing, rotation_rate=Ω_Earth, rotation_axis=nothing, latitude=nothing)

    if latitude !=nothing
        isnothing(rotation_axis) && throw(ArgumentError("Cannot specify latitude and rotation axis."))
        all(isnothing.((fx, fy, fz)) || throw(ArgumentError("Cannot specify latitude and (fx, fy, fz)."))
        # calculate rotation axis
    end

    if rotation_axis != nothing
        all(isnothing.((fx, fy, fz)) || throw(ArgumentError("Cannot specify rotation_axis and (fx, fy, fz)."))
        # calculate fx, fy, fz
    end

    fx, fy, fz = zero_if_nothing.((fx, fy, fz)) # set default fx, fy, fz

    return ConstantBackgroundRotation(FT(fx), FT(fy), FT(fz))
end
```

There's also the possibility of a somewhat minor optimization by keeping the possibility that `fx, fy, fz` might be `nothing`, and eliding the associated Coriolis operations in kernel functions for that case. This requires the struct to accomodate different types for each of them.

tomchor commented:
> For API this is what I suggest:
> 
> Three "modes":
> 
> 1. A "primitive" mode whereby fx, fy, fz are all explicitly provided.
> 2. A "general rotation" mode whereby `rotation_axis` and `rotation_rate` are provided. `fx`, `fy`, and `fz` are then calculated as a convenience.
> 3. A "tangent plane" mode whereby `latitude` and `rotation_rate` are provided. This mode calculates `rotation_axis` based on `latitude` (and assuming that `y` is north-south).

Thanks for the suggestion. That was kind of the API for `NonTradFPlane`. I actually think three modes is kinda overkill and it makes the constructor function unnecessarily complicated (which is why I simplified it this time), especially since it's really not hard to calculate one mode from another one. I'd suggest keeping two of them as a compromise: "general rotation" (to match the API in `Buoyancy`) and "tangent plane". What do you think?


> I think `GeneralFPlane` is ambiguous, so I propose using a more specific name.

I completely agree. I used `GeneralFPlane` more as placeholder. But as you pointed out in #1372, `ConstantBackgroundRotation` is also kinda ambiguous since every rotation we have is technically a constant background rotation at its core. How about get descriptive and call it `ArbitraryAxisFPlane`?

glwagner commented:
> > For API this is what I suggest:
> > Three "modes":
> > 
> > 1. A "primitive" mode whereby fx, fy, fz are all explicitly provided.
> > 2. A "general rotation" mode whereby `rotation_axis` and `rotation_rate` are provided. `fx`, `fy`, and `fz` are then calculated as a convenience.
> > 3. A "tangent plane" mode whereby `latitude` and `rotation_rate` are provided. This mode calculates `rotation_axis` based on `latitude` (and assuming that `y` is north-south).
> 
> Thanks for the suggestion. That was kind of the API for `NonTradFPlane`. I actually think three modes is kinda overkill and it makes the constructor function unnecessarily complicated (which is why I simplified it this time), especially since it's really not hard to calculate one mode from another one. I'd suggest keeping two of them as a compromise: "general rotation" (to match the API in `Buoyancy`) and "tangent plane". What do you think?
> 
> > I think `GeneralFPlane` is ambiguous, so I propose using a more specific name.
> 
> I completely agree. I used `GeneralFPlane` more as placeholder. But as you pointed out in #1372, `ConstantBackgroundRotation` is also kinda ambiguous since every rotation we have is technically a constant background rotation at its core. How about get descriptive and call it `ArbitraryAxisFPlane`?

I made my suggestion because I disagree that supporting those keyword arguments is overkill. I am arguing that the complexity of the constructor is justified because

1) not supporting this API moves part of the "complexity" from the constructor into user scripts, which is very undesirable in general
2) having a consistent API with `FPlane` and `BetaPlane` (which both support a `latitude` keyword argument) is important
3) I think it really is a useful and friendly simplification to abstract the projection of the rotation axis into a plane tangent with a spherical surface (I would personally benefit from this as a user, which I feel implies that many other users will also benefit).

As for a name, maybe `RectilinearCoriolis? `BetaPlane` is specific to rectilinear grids but isn't a constant background rotation rate; `HydrostaticSphericalCoriolis` is for spherical shells. `FPlane` overlaps but we conceded that point because it's so common.

francispoulin commented:
I vote for the 3 "modes" as well, as it is similar to the beta-plane and it does give the user more options.

tomchor commented:
> I made my suggestion because I disagree that supporting those keyword arguments is overkill. I am arguing that the complexity of the constructor is justified because
> 
> 1. not supporting this API moves part of the "complexity" from the constructor into user scripts, which is very undesirable in general
> 2. having a consistent API with `FPlane` and `BetaPlane` (which both support a `latitude` keyword argument) is important
> 3. I think it really is a useful and friendly simplification to abstract the projection of the rotation axis into a plane tangent with a spherical surface (I have personally benefited from this abstraction for setting up examples like the one proposed in [Overturning, eddying channel example following Abernathey et al. 2011 #1744](https://github.com/CliMA/Oceananigans.jl/pull/1744), which I feel implies that many other users will also benefit).

Fair enough. I was leaning towards simplicity as per our last conversation about simplifying the code to make it easier for development and (mostly) because I think a sine/cosine multiplication doesn't add much complexity on the user side. But it seems I'm voted out so I'll implement those functionalities back :)

tomchor commented:
> As for a name, maybe `RectilinearCoriolis`? This contrasts with `BetaPlane`, which is specific to rectilinear grids but isn't a constant background rotation rate, and `HydrostaticSphericalCoriolis`, which is for hydrostatic fluids in spherical shells. `FPlane` overlaps but we can concede this point because it's so common (and perhaps discuss eliminating `FPlane` in a future PR).

The thing is that `BetaPlane` is also a "rectilinear Coriolis" since it's a Coriolis implementation meant for rectilinear grids, no? I'm not very familiar with the implementation of the Coriolis acc in spherical grids, but if I understand correctly you can't use `FPlane` or `BetaPlane` on spherical grids?

If that's true, maybe we aim for consistency with the rectilinear grids for now and rename things to

- `RectilinearGridFPlane` (this one)
- `RectilinearGridBetaPlane` (what is now `BetaPlane`)
- Keep `FPlane` for now because, like you said, it's an easy name that's very commonly used

I do think that it's important to keep "f-plane" in the name of this implementation somewhere instead of just "Coriolis" because there are different approximations one can make when implementing the Coriolis acc, so it'd help to be specific.

glwagner commented:
> > As for a name, maybe `RectilinearCoriolis`? This contrasts with `BetaPlane`, which is specific to rectilinear grids but isn't a constant background rotation rate, and `HydrostaticSphericalCoriolis`, which is for hydrostatic fluids in spherical shells. `FPlane` overlaps but we can concede this point because it's so common (and perhaps discuss eliminating `FPlane` in a future PR).
> 
> The thing is that `BetaPlane` is also a "rectilinear Coriolis" since it's a Coriolis implementation meant for rectilinear grids, no? I'm not very familiar with the implementation of the Coriolis acc in spherical grids, but if I understand correctly you can't use `FPlane` or `BetaPlane` on spherical grids?
> 
> If that's true, maybe we aim for consistency with the rectilinear grids for now and rename things to
> 
> * `RectilinearGridFPlane` (this one)
> * `RectilinearGridBetaPlane` (what is now `BetaPlane`)
> * Keep `FPlane` for now because, like you said, it's an easy name that's very commonly used
> 
> I do think that it's important to keep "f-plane" in the name of this implementation somewhere instead of just "Coriolis" because there are different approximations one can make when implementing the Coriolis acc, so it'd help to be specific.

The f-plane refers to a tangent plane approximation for a fluid contained in a spherical shell. However, this type is more general than that case (this was also a problem with the undecorated `FPlane` type), and can be used to model any rotating fluid in a rectilinear domain with an arbitrary axis of rotation. Therefore, I argue not using `FPlane` in the name of this type.

`BetaPlane` has no meaning outside of rectilinear geometries, therefore I propose not adding the superfluous `RectilinearGrid` in front of `BetaPlane`.

As for `Grid` or not, the meaning of `RectilinearGrid` is specific to Oceananigans. However, the term "rectilinear" has a distinct meaning apart from "grids". Here, I think its clear if the name refers to the coordinate system (rather than the grid or mesh), I think I prefer simply "RectilinearCoriolis". Otherwise, we will also need to consider how to devise a similar name for "HydrostaticSphericalCoriolis" that somehow references the handful of grid types that we permit for in spherical domains.

glwagner commented:
Well, I have to say I've been confused about the distinction between "Cartesian" and "Rectilinear" for a while. Maybe this is showing here.

I think the system we are developing (somewhat implicitly, but we will need to formalize it eventually) is that every "grid" is attached to a coordinate system. Therefore a "rectilinear" grid has a Cartesian coordinate system, with the three coordinates x, y, z.

In contrast a "spherical" grid uses the geographic coordinate system (latitude, longitude, altitude).

In either case, the grid may not actually coincide with coordinate lines. For example, we can have a curvilinear grid in a Cartesian geometry. However, almost all the grids (except on the cubed sphere) we have exhibit this property.

I think the discussion we are having is really specific to the coordinate system rather than the grid. We need to have different abstractions for Coriolis forces for different coordinate systems, because the data we need to store is different. For a Cartesian coordinate system, we'd like to keep track of the Cartesian components of the Coriolis vector (fx, fy, fz). For a geographic spherical coordinate system we probably want to express the latitude pierced by the rotation axis (in our current abstraction we only support a rotation axis through the pole).

Long story short... perhaps something like `ConstantCartesianCoriolis`?

tomchor commented:
I see your points about this being actually more general than an f-plane approximation, and I agree that the discussion about the grids is a difficult one that is taking over this issue.

Before you sent this last comment I was gonna suggest `RectilinearConstantCoriolis`. But after your points being just `ConstantCoriolis`? And then the "cartesian"/"rectilinear" is implicit, just like it's the case with `FPlane`? If not, I'm okay with calling it `ConstantCartesianCoriolis` like you suggested.

Maybe after merging this PR we can open an issue to discuss the Coriolis/grids issues more in depth?

francispoulin commented:
My two cents.

Both the f-plane and beta-plane approximations approximate the Coriolis pseudo-force from a spherical coordiante, using latitude \varphi, to a planar geometry using typicallay y and z.  Since they are both approximations into a plane (which could be tilted as this PR is trying to do) how about `ConstantPlanarCoriolis` for `f-plane` and `LinearPlanarCoriolis` for `beta-plane`? 

In Calculus, these approximations can be thought of as a constant approximation (`f-plane`) or linear approximation (`beta-plane`).  Another idea is to call them simply `ConstantCoriolis` and `LinearCoriois`.  

I agree that `Grid` is not really relevant since we always use a grid.

Also, `Rectilinear` does not seem as descriptive as

francispoulin commented:
It seemed that @tomchor and I both came up with the idea of `ConstantCoriolis` within a short span of time.  It's simple and accurate.

tomchor commented:
Also, a quick question that might be silly, if I may. Why isn't the Earth's rotation rate 2π rad / 24hours? It's very close but not exactly there:

```julia
julia> 2π/24hours ≈ Oceananigans.Coriolis.Ω_Earth
false

julia> Oceananigans.Coriolis.Ω_Earth
7.292115e-5

julia> 2π/24hours
7.27220521664304e-5
```

glwagner commented:
> Also, a quick question that might be silly, if I may. Why isn't the Earth's rotation rate 2π rad / 24hours? It's very close but not exactly there:
> 
> ```julia
> julia> 2π/24hours ≈ Oceananigans.Coriolis.Ω_Earth
> false
> 
> julia> Oceananigans.Coriolis.Ω_Earth
> 7.292115e-5
> 
> julia> 2π/24hours
> 7.27220521664304e-5
> ```

Because 24 hours is not the rotation period of the Earth! Though it is pretty close, which is one reason why calendars are fairly good (but not perfect)...

https://en.wikipedia.org/wiki/Earth%27s_rotation#Stellar_day

tomchor commented:
> Because 24 hours is not the rotation period of the Earth! Though it is pretty close...
> 
> https://en.wikipedia.org/wiki/Earth%27s_rotation#Stellar_day

Ah, forgot about that fact haha. Although I always thought the difference was smaller...

glwagner commented:
Is it possible to calculate the components `fx, fy, fz` on the fly? Also, can we infer `rotation_axis` from user-specified `fx, fy, fz`? If so, we can have one type for a constant rotation rate (storing just the parameters `rotation_axis` and `rotation_rate`), and use it for both Cartesian and spherical geometries.

glwagner commented:
> Is it possible to calculate the components `fx, fy, fz` on the fly? Also, can we infer `rotation_axis` from user-specified `fx, fy, fz`? If so, we can have one type for a constant rotation rate (storing just the parameters `rotation_axis` and `rotation_rate`), and use it for both Cartesian / rectilinear and spherical geometries.

EDIT: this might be a bad idea if it adds functions like sin / cos to kernels.

If we do need to pre-calculate fx, fy, fz, I think we should somehow indicate that this type is specifically Cartesian or rectilinear to avoid confusion with its spherical-geometry counterpart... unfortuntately, because its annoying.

tomchor commented:
> Is it possible to calculate the components `fx, fy, fz` on the fly? Also, can we infer `rotation_axis` from user-specified `fx, fy, fz`? If so, we can have one type for a constant rotation rate (storing just the parameters `rotation_axis` and `rotation_rate`), and use it for both Cartesian / rectilinear and spherical geometries.

We definitely can do those things, I'm just not sure how it'll work with kernels. Unless I'm missing something:

```julia
fx = 2rotation_rate * rotation_axis[1]
fy = 2rotation_rate * rotation_axis[2]
fz = 2rotation_rate * rotation_axis[3]
```

and then to get `rotation_axis`:

```julia
rotation_axis = [fx, fy, fz] ./ √(fx^2 + fy^2 + fz^2)
```

This would be kind of a big change though, no? Is it worth doing it in this PR?

Also, we'd need to make the difference between rotation rate and Coriolis frequency clear (meaning `f = 2rotation_rate`).

glwagner commented:
> > Is it possible to calculate the components `fx, fy, fz` on the fly? Also, can we infer `rotation_axis` from user-specified `fx, fy, fz`? If so, we can have one type for a constant rotation rate (storing just the parameters `rotation_axis` and `rotation_rate`), and use it for both Cartesian / rectilinear and spherical geometries.
> 
> We definitely can do those things, I'm just not sure how it'll work with kernels. Unless I'm missing something:
> 
> ```julia
> fx = 2rotation_rate * rotation_axis[1]
> fy = 2rotation_rate * rotation_axis[2]
> fz = 2rotation_rate * rotation_axis[3]
> ```
> 
> and then to get `rotation_axis`:
> 
> ```julia
> rotation_axis = [fx, fy, fz] ./ √(fx^2 + fy^2 + fz^2)
> ```
> 
> This would be kind of a big change though, no? Is it worth doing it in this PR?
> 
> Also, we'd need to make the difference between rotation rate and Coriolis frequency clear (meaning `f = 2rotation_rate`).

Well, that would be just fine for kernels (only change is that we need to use a tuple rather than array, so `(fx, fy, fz)` rather than `[fx, fy, fz]`).

But again I think about it a bit more and one annoyance is that we need a version that makes the hydrostatic approximation. *sigh* so maybe we are back to something cumbersome like `ConstantCartesianCoriolis`...

tomchor commented:
> But again I think about it a bit more and one annoyance is that we need a version that makes the hydrostatic approximation. _sigh_ so maybe we are back to something cumbersome like `ConstantCartesianCoriolis`...

By hydrostatic approximation do you mean that ignores `fx` and `fy`? Because f so, I don't think that's a problem. We could set something that works like this PR is working now:

```julia
full_coriolis = ConstantCoriolis(latitude=45) # and then rotation rate and axis are calculated in full
vert_coriolis = ConstantCoriolis(f=2*sind(45)*Ω, rotation_axis=[0, 0, 1])
```

glwagner commented:
> > But again I think about it a bit more and one annoyance is that we need a version that makes the hydrostatic approximation. _sigh_ so maybe we are back to something cumbersome like `ConstantCartesianCoriolis`...
> 
> 
> 
> By hydrostatic approximation do you mean that ignores `fx` and `fy`? Because f so, I don't think that's a problem. We could set something that works like this PR is working now:
> 
> 
> 
> ```julia
> 
> full_coriolis = ConstantCoriolis(latitude=45) # and then rotation rate and axis are calculated in full
> 
> vert_coriolis = ConstantCoriolis(f=2*sind(45)*Ω, rotation_axis=[0, 0, 1])
> 
> ```

I'm talking about the hydrostatic approximation in a spherical coordinate system, where do not use the Cartesian components.

tomchor commented:
> I'm talking about the hydrostatic approximation in a spherical coordinate system, where do not use the Cartesian components.

Ah, I see. I'm not sure how that would factor in since I'm not familiar with the equations. But I'll trust that for now it's back to a separate constructor as we've been talking about.

tomchor commented:
If all the tests pass then this PR is ready for its final review. Notes:

- I implemented the 3-input-modes functionality that @glwagner suggested and added tests to make sure it's working properly
- @francispoulin and I independently came up with the name `ConstantCoriolis` so I went with that name for now (as opposed to the a-bit-more-obscure `ConstantCartesianCoriolis`). Let me know if anyone objects and it should be easy to change it to something else.
- Changed the docs accordingly in model_setup/coriolis and physics/coriolis (also fixed a couple of typos), but let me know if there's somewhere else in the docs where this should be changed

glwagner commented on docs/src/model_setup/coriolis.md:
```diff
@@ -38,23 +38,36 @@ FPlane{Float64}: f = 1.03e-04
 in which case the value of ``f`` is given by ``2\Omega\sin\varphi``.
 
-## Non-traditional ``f``-plane
+## Constant Coriolis
```
```suggestion
## Coriolis term for constant rotation in a Cartesian coordinate system
```

glwagner commented on src/Coriolis/constant_coriolis.jl:
```diff
@@ -0,0 +1,82 @@
+rotation can be specified in three different ways:
+
+- Specifying all components `fx`, `fy` and `fz` directly.
+- Specifying the Coriolis frequency `f` and (optionally) a `rotation_axis` (which defaults to the
```
Might be less controversial to call it the Coriolis "parameter" --- most people interpret `f` as a vorticity (eg the planetary vorticity, or the vorticity associated with solid body rotation); even though it also sets the time-scale for inertial waves that are perturbations to solid body rotation.

tomchor replied:
Done

glwagner commented:
Ah sorry I really do think it would be confusing to users that one has to use `HydrostaticSphericalCoriolis` on grids that have spherical coordinate systems :-( don't others agree?

francispoulin commented:
This is a bit complicated but it is better to have discussion now to plan what we think is best moving forward.

It does seem that we want to make sure that we can't use f- and beta-planes in a spherical geometry, and don't want to use the spherical Coriolis force in a Cartesian geometry. This can be done in a variety of ways I suppose.  If adding Cartesian in the name helps for the moment then why not?  That way we can merge sooner rather than later and then discuss this as an issue.

We all agree that we need different Coriolis functions for Cartesian and Spherical coordinates.  Why do we need different functions or nonhydrostatic (and shallow water I presumme) and hydrostatic?

tomchor commented:
> Ah sorry I really do think it would be confusing to users that one has to use HydrostaticSphericalCoriolis on grids that have spherical coordinate systems; and that the "constant Coriolis" type only applies to Cartesian coordinate systems :-( don't others agree?

I do agree it's a bit confusing, but IMO it's also obscure to name it `ConstantCartesianCoriolis`. As a user I'm thinking what's a cartesian Coriolis? And I definitely wouldn't make the leap in reasoning that `ConstantCartesianCoriolis` is meant to be used with _rectilinear grids_ (none of those two words appears in the name). Which is I suggested a few posts back to use the (very verbose) `ConstantRectlinearGridCoriolis` or something to that effect.

> Sorry this is becoming laborious... if the name is changed to include Cartesian then we can merge this and discuss further in an issue.

I'm okay with this. It seems like the background rotation implementation could use a big refactoring, which is kind of outside the scope of this PR anyway.

tomchor commented:
> We all agree that we need different Coriolis functions for Cartesian and Spherical coordinates. Why do we need different functions or nonhydrostatic (and shallow water I presumme) and hydrostatic?

I'm also not clear, although I'm assuming that it has to do with `w` being treated differently (since `w` appears in the equations for the horizontal rotation components). @glwagner can you clarify?

glwagner commented:
> > We all agree that we need different Coriolis functions for Cartesian and Spherical coordinates. Why do we need different functions or nonhydrostatic (and shallow water I presumme) and hydrostatic?
> 
> I'm also not clear, although I'm assuming that it has to do with `w` being treated differently (since `w` appears in the equations for the horizontal rotation components). @glwagner can you clarify?

Yes, when we make the hydrostatic approximation, we assume that the aspect ratio is thin and H/L is small. The hydrostatic assumption specifically refers to the use of this scaling in the vertical momentum equation (reducing it to a diagnostic equation for hydrostatic pressure). This same scaling applied to the Coriolis force leads to the "traditional" approximation such that Coriolis terms involving the _vertical_ velocity are neglected from the horizontal momentum equations (likewise, the terms involving the horizontal momentum are neglected from the vertical momentum balance). The traditional approximation is also used for the "f-plane" and "beta-plane" when the domain is supposed to approximate oceanic motion away from the poles. In Oceananigans, we have to express this notion with a _type_ to avoid adding spurious terms to the horizontal momentum equations that depend on the vertical velocity.

francispoulin commented:
@glwagner : I understand what you are saying about the traditional approximation.  

If we include the non-tradidtional terms, we obtain the Quasi-Hydrostatic model, as discussed in the MITgcm.  This is something we can solve in the `HydrostaticModel`, right? 

I'm not suggesting we change the name of the model but I thought we could use this in either model, but maybe there's a problem because of the pressure solve, which might not allow for this?

glwagner commented:
> @glwagner : I understand what you are saying about the traditional approximation.
> 
> If we include the non-tradidtional terms, we obtain the Quasi-Hydrostatic model, as discussed in the MITgcm. This is something we can solve in the `HydrostaticModel`, right?
> 
> I'm not suggesting we change the name of the model but I thought we could use this in either model, but maybe there's a problem because of the pressure solve, which might not allow for this?

We could indeed implement such a model; it would not be very difficult. It would mean that the hydrostatic pressure depends on the `coriolis` type.

Such a modification to the hydrostatic pressure integral is also needed to introduce surface wave effects via the Craik-Leibovich approximation to a hydrostatic model.

francispoulin commented:
> We could indeed implement such a model; it would not be very difficult. It would mean that the hydrostatic pressure depends on the `coriolis` type.
> 
> Such a modification to the hydrostatic pressure integral is also needed to introduce surface wave effects via the Craik-Leibovich approximation to a hydrostatic model (this effects can be interpreted as a modification to the background rotation rate of the fluid, with the vertical derivative of the Stokes drift acting as the Coriolis vector).

Okay, so a longer term plan.

For the moment we need that `HydrostaticModel` uses traditional f- or beta-plane, or something that does the same thing.  It is `NonhydrostaticModel` that will allow for more types of coriolis terms.  Understood.

glwagner commented:
> > We could indeed implement such a model; it would not be very difficult. It would mean that the hydrostatic pressure depends on the `coriolis` type.
> > Such a modification to the hydrostatic pressure integral is also needed to introduce surface wave effects via the Craik-Leibovich approximation to a hydrostatic model (this effects can be interpreted as a modification to the background rotation rate of the fluid, with the vertical derivative of the Stokes drift acting as the Coriolis vector).
> 
> Okay, so a longer term plan.
> 
> For the moment we need that `HydrostaticModel` uses traditional f- or beta-plane, or something that does the same thing. It is `NonhydrostaticModel` that will allow for more types of coriolis terms. Understood.

Note too that the problem is really specific to a spherical geometry. In a rectangular geometry, approximations regarding the Coriolis vector can be expressed by changing the rotation axis. In a spherical geometry this is no longer possible and we need to explicitly state that we are making a hydrostatic approximation.

tomchor commented:
Added some dynamics tests for Coriolis with this last commit. It runs a 0-D case for half an inertial period with a rotation in the `z` axis and `x` axes and then compares both to make sure they produce the same result (but rotated).

There's one part that tests if the total velocity magnitude is approximately unchanged (magnitude=1), which relies on an implicit arbitrary tolerance which might be bad. I'd curious about your feedback on that one.

>  if the name is changed to include Cartesian then we can merge this and discuss further in an issue.

Per the comment above I'm going to change the name to `ConstantCartesianCoriolis` and (provided the tests all pass and you're okay with my new test addition) I'll proceed to merge this into master and open an issue to further discuss the issues that emerged here.

glwagner commented:
> Added some dynamics tests for Coriolis with this last commit. It two a 0-D case for half an inertial period with a rotation about the `z` axis and `x` axis and then compares both to make sure they produce the same result (but rotated).
> 
> There's one part that tests if the total velocity magnitude is approximately unchanged (magnitude=1), which relies on an implicit arbitrary tolerance which might be bad. I'd curious about your feedback on that one.
> 
> > if the name is changed to include Cartesian then we can merge this and discuss further in an issue.
> 
> Per the comment above I'm going to change the name to `ConstantCartesianCoriolis` and (provided the tests all pass and you're okay with my new test addition) I'll proceed to merge this into master and open an issue to further discuss the issues that emerged here.

For physics tests we can't avoid introducing an arbitrary tolerance. So its ok. That's one reason why physics tests in CI are a bit problematic and we also need validation tests analyzed by humans.

glwagner reviewed (approved):
Thanks @tomchor for the hard work and enduring a long discussion about semantics.

tomchor commented:
> Thanks @tomchor for the hard work and enduring a long discussion about semantics.

Np. This was a good illustration of scientific debate haha
